### PR TITLE
Fix #8516: Add Open in Brave to iOS Share Sheet Action (Action Extension)

### LINKED ISSUE #8516: Add Open in Brave to iOS Share Sheet Action (Action Extension)
## Problem Description 

 Brave Browser should appear in OS default Share Sheet for URL related actions. This will provide extra visibility and easy access to Brave users on iOS thorough other applications while using the Share Sheet. 

## Feature Overview 

Adding Brave as Action Extension on iOS will add `Open in Brave` action to default share sheet. So Brave can be reached outside the application environment. It will also help for users having multiple browsers and want to use Brave for certain actions or links.

## Design 

There is no design choice to be made here but it will look like the actions shown in he example

## Implementation Details 

Action extensions are displayed in the action view controller, accessed by users when they tap the Share button within an active application. These extensions are situated in the action section of this panel, alongside default actions like printing and copying content.

When an Action extension is developed, it needs to specify the content types it is capable of handling.

For Brave's case this should be 

NSExtensionActivationSupportsText
NSExtensionActivationSupportsWebPageWithMaxCount
NSExtensionActivationSupportsWebURLWithMaxCount

- An action extension target should be added
- Display Name of Extension should be changed
- ActionViewController should be implemented with item provider conforming identifier type and load

## User Experience 

1. Open any app and share url or text
2. Check share sheet has open in brave action
3. this share functionality should search the text in brave or open link

### Additional information 

It needs be tested and implemented to see if this behaviour could be possible to detect it is a text or link and do desired action according to it.

### PR DESCRIPTION
**Pull request by soner-yuksel**
## Summary of Changes

This PR is enabling Brave Browser OS default Share Sheet for URL related actions. This will provide extra visibility and easy access to Brave users on iOS thorough other applications while using the Share Sheet.

This pull request fixes #8516

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:

- Open any app like Safari
- Highlight a text or a url
- Share
- Check Open in Brave action is there 
- Test using the open in brave actions for various types

## Screenshots:

https://github.com/brave/brave-ios/assets/6643505/b89d26ed-8143-439e-9d08-72d7349ae6e8

Screenshot 2024-01-19 at 2 55 19 PM

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: App/Client.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -2442,6 +2567,306 @@
...
+				DEBUG_INFORMATION_FORMAT = dwarf;
+				ENABLE_TESTABILITY = YES;
+				ENABLE_USER_SCRIPT_SANDBOXING = YES;
+				GCC_C_LANGUAGE_STANDARD = gnu17;
+				GCC_PREPROCESSOR_DEFINITIONS = "DEBUG=1";
+				GENERATE_INFOPLIST_FILE = NO;
+				INFOPLIST_FILE = ActionExtension/ActionExtension.plist;
+				INFOPLIST_KEY_CFBundleDisplayName = "";
+				INFOPLIST_KEY_NSHumanReadableCopyright = "";
+				IPHONEOS_DEPLOYMENT_TARGET = 15.0;
```
Comment (kylehickinson) [inline App/Client.xcodeproj/project.pbxproj]: Delete this override so it follows project settings

File: App/Client.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -2442,6 +2567,306 @@
...
+				LD_RUNPATH_SEARCH_PATHS = (
+					"$(inherited)",
+					"@executable_path/Frameworks",
+					"@executable_path/../../Frameworks",
+				);
+				LOCALIZATION_PREFERS_STRING_CATALOGS = YES;
```
Comment (kylehickinson) [inline App/Client.xcodeproj/project.pbxproj]: Delete this override so it follows project settings

File: App/Client.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -2442,6 +2567,306 @@
...
+				MARKETING_VERSION = 1.0;
+				MTL_ENABLE_DEBUG_INFO = INCLUDE_SOURCE;
+				MTL_FAST_MATH = YES;
+				PRODUCT_BUNDLE_IDENTIFIER = "$(inherited).$(PRODUCT_NAME:rfc1034identifier)";
+				PRODUCT_NAME = "$(TARGET_NAME)";
+				SKIP_INSTALL = YES;
+				SWIFT_ACTIVE_COMPILATION_CONDITIONS = "DEBUG $(inherited)";
+				SWIFT_EMIT_LOC_STRINGS = NO;
```
Comment (kylehickinson) [inline App/Client.xcodeproj/project.pbxproj]: Delete this override so it follows project settings

File: App/Client.xcodeproj/project.pbxproj (hunk 1 above)
Comment (soner-yuksel) [reply App/Client.xcodeproj/project.pbxproj]: Thanks for the warning. It is removed with https://github.com/brave/brave-ios/pull/8682/commits/7202abb15df5e102b478290989958c8d69fee1a5

File: App/Client.xcodeproj/project.pbxproj (hunk 2 above)
Comment (soner-yuksel) [reply App/Client.xcodeproj/project.pbxproj]: Removed at https://github.com/brave/brave-ios/pull/8682/commits/7202abb15df5e102b478290989958c8d69fee1a5

File: App/Client.xcodeproj/project.pbxproj (hunk 3 above)
Comment (soner-yuksel) [reply App/Client.xcodeproj/project.pbxproj]: Removed at https://github.com/brave/brave-ios/pull/8682/commits/7202abb15df5e102b478290989958c8d69fee1a5

File: App/ActionExtension/ActionToBraveViewController.swift [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+          loadAttachmentFor(type: .url, using: provider)
+          break
+        }
+      }
+    }
+  }
+  
+  func done() {
+    // Return any edited content to the host app, in this case empty
+    extensionContext!.completeRequest(returningItems: [], completionHandler: nil)
```
Comment (iccub) [inline App/ActionExtension/ActionToBraveViewController.swift]: do we have to force unwrap here?

File: App/ActionExtension/ActionToBraveViewController.swift [hunk 5]
```diff
@@ -0,0 +1,118 @@
+// Copyright 2023 The Brave Authors. All rights reserved.
```
Comment (iccub) [inline App/ActionExtension/ActionToBraveViewController.swift]: 2024 old man!

File: App/ActionExtension/ActionToBraveViewController.swift [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+      
+      return createURL(for: .url, with: queryURL)
+    }
+    
+  }
+  
+  private func createURL(for schemeType: SchemeType, with value: String) -> URL? {
+    var queryItem: URLQueryItem
+    var components = URLComponents()
+    components.scheme = Bundle.main.infoDictionary?["BRAVE_URL_SCHEME"] as? String ?? "brave"
```
Comment (iccub) [inline App/ActionExtension/ActionToBraveViewController.swift:86]: `BRAVE_URL_SCHEME`  what is this for?

**Review by iccub — CHANGES_REQUESTED** [review]

File: App/ActionExtension/ActionToBraveViewController.swift (hunk 6 above)
Comment (soner-yuksel) [reply App/ActionExtension/ActionToBraveViewController.swift:86]: Similar functionality like in ShareExtensions creating the deep link  with brave scheme and the url to be opened.

File: App/ActionExtension/ActionToBraveViewController.swift (hunk 4 above)
Comment (soner-yuksel) [reply App/ActionExtension/ActionToBraveViewController.swift]: No certainly not, good catch. 

Fixed at https://github.com/brave/brave-ios/pull/8682/commits/640b3481565a6f54ab6bf9830aa289d1a26d5d68

File: App/ActionExtension/ActionToBraveViewController.swift (hunk 5 above)
Comment (soner-yuksel) [reply App/ActionExtension/ActionToBraveViewController.swift]: Changed https://github.com/brave/brave-ios/pull/8682/commits/640b3481565a6f54ab6bf9830aa289d1a26d5d68

**Review by iccub — APPROVED** [review]
more code changes than I anticipated. nice job

**Pull request merged**